src/messaging/ExchangeContext.cpp

```diff
@@ -596,21 +596,26 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
     // Set kFlagReceivedAtLeastOneMessage to true since we have received at least one new application level message
     SetHasReceivedAtLeastOneMessage(true);
 
-    if (IsResponseExpected())
+    // Don't send messages on to our delegate if our dispatch does not allow
+    // those messages.  Those messages should also not be treated as responses,
+    // since if our delegate is expecting a response we will not notify it about
+    // these messages.
+    if (mDispatch.MessagePermitted(payloadHeader.GetProtocolID(), payloadHeader.GetMessageType()))
```

Comment:
Is this change something we can unit test? It would be great if we could ensure we never regress again on this.

Reply:
Discussed offline, we can add the test as we work on test infra, so it will be a separate PR.

Reply:
It should be, but will take some work.  I'll do the test as a followup.